tool/tsh/gcp.go

```diff
@@ -0,0 +1,443 @@
+	// 	return trace.Wrap(err)
+	// }
+
+	a.localALPNProxy, err = alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
+		Listener:           listener,
+		RemoteProxyAddr:    tc.WebProxyAddr,
+		Protocols:          []alpncommon.Protocol{alpncommon.ProtocolHTTP},
+		InsecureSkipVerify: a.cf.InsecureSkipVerify,
+		ParentContext:      a.cf.Context,
+		SNI:                address.Host(),
+		Certs:              []tls.Certificate{appCerts},
+		HTTPMiddleware: &alpnproxy.GCPMiddleware{
+			Secret: a.secret,
+		},
+	})
+
+	if err != nil {
+		if cerr := listener.Close(); cerr != nil {
+			return trace.NewAggregate(err, cerr)
+		}
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		if err := a.localALPNProxy.StartHTTPAccessProxy(a.cf.Context); err != nil {
+			log.WithError(err).Errorf("Failed to start local ALPN proxy.")
+		}
+	}()
+	return nil
+}
+
+// startLocalForwardProxy starts the local forward proxy.
+func (a *gcpApp) startLocalForwardProxy(port string) error {
+	listenAddr := "localhost:0"
+	if port != "" {
+		listenAddr = fmt.Sprintf("localhost:%s", port)
+	}
+
+	// Note that the created forward proxy serves HTTP instead of HTTPS, to
+	// eliminate the need to install temporary CA for various GCP clients.
+	listener, err := net.Listen("tcp", listenAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	a.localForwardProxy, err = alpnproxy.NewForwardProxy(alpnproxy.ForwardProxyConfig{
+		Listener:     listener,
+		CloseContext: a.cf.Context,
+		Handlers: []alpnproxy.ConnectRequestHandler{
+			// Forward GCP requests to ALPN proxy.
+			alpnproxy.NewForwardToHostHandler(alpnproxy.ForwardToHostHandlerConfig{
+				MatchFunc: alpnproxy.MatchGCPRequests,
+				Host:      a.localALPNProxy.GetAddr(),
+			}),
+
+			// Forward non-GCP requests to user's system proxy, if configured.
+			alpnproxy.NewForwardToSystemProxyHandler(alpnproxy.ForwardToSystemProxyHandlerConfig{
+				InsecureSystemProxy: a.cf.InsecureSkipVerify,
+			}),
+
+			// Forward non-GCP requests to their original hosts.
+			alpnproxy.NewForwardToOriginalHostHandler(),
+		},
+	})
+	if err != nil {
+		if cerr := listener.Close(); cerr != nil {
+			return trace.NewAggregate(err, cerr)
+		}
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		if err := a.localForwardProxy.Start(); err != nil {
+			log.WithError(err).Errorf("Failed to start local forward proxy.")
+		}
+	}()
+	return nil
+}
+
+func printGCPServiceAccounts(accounts []string) {
+	fmt.Println(formatGCPServiceAccounts(accounts))
+}
+
+// SortedGCPServiceAccounts sorts service accounts by project and service account name.
+type SortedGCPServiceAccounts []string
+
+// Len returns the length of a list.
+func (s SortedGCPServiceAccounts) Len() int {
+	return len(s)
+}
+
+// Less compares items by name.
+func (s SortedGCPServiceAccounts) Less(i, j int) bool {
+	beforeI, afterI, _ := strings.Cut(s[i], "@")
+	beforeJ, afterJ, _ := strings.Cut(s[j], "@")
+
+	if afterI != afterJ {
+		return afterI < afterJ
+	}
+
+	return beforeI < beforeJ
+}
```

Comment:
Could you add docs describing the expected behavior of this sorting? I think it's sorting first by what follows @ and then by what precedes "@" if the suffix is equal?

Reply:
Good point, will do.